mmtools/web_search.go

```diff
@@ -0,0 +1,951 @@
+			return "you can only fetch URLs that were returned from web search results", errors.New("no whitelist in context")
+		}
+	}
+
+	// Security check 2: Check if the domain is blacklisted
+	cfg := s.cfgGetter()
+	if cfg != nil && isBlacklisted(pageURL, cfg.WebSearch.DomainBlacklist) {
+		s.logWarn("source fetch blocked by domain blacklist", "url", pageURL)
+		return "this domain is blocked by the administrator's configuration", errors.New("domain blacklisted")
+	}
+
+	client := s.httpClient
+	if client == nil {
+		client = http.DefaultClient
+	}
+
+	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, pageURL, nil)
+	if err != nil {
+		s.logError("failed to create source fetch request", "error", err)
+		return "unable to create request", err
+	}
+	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8")
+	req.Header.Set("User-Agent", "Mattermost-AI-Plugin/1.0")
+
+	resp, err := client.Do(req)
+	if err != nil {
+		s.logError("source fetch request failed", "error", err, "url", pageURL)
+		return "unable to fetch the requested URL", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode < http.StatusOK || resp.StatusCode >= http.StatusMultipleChoices {
+		s.logWarn("source fetch non-success status", "status", resp.Status, "url", pageURL)
+		return fmt.Sprintf("failed to fetch URL: %s", resp.Status), fmt.Errorf("source fetch failed: %s", resp.Status)
+	}
+
+	// Limit the read size to prevent DoS (e.g., 30MB limit)
+	const maxDownloadSize = 30 * 1024 * 1024
+	body, err := io.ReadAll(io.LimitReader(resp.Body, maxDownloadSize))
+	if err != nil {
+		s.logError("failed to read source fetch response", "error", err, "url", pageURL)
+		return "unable to read the response", err
+	}
+
+	// Extract content
+	var textContent string
+	parsedURL, parseErr := url.Parse(pageURL)
+	if parseErr == nil {
+		article, readErr := readability.FromReader(bytes.NewReader(body), parsedURL)
+		if readErr == nil && strings.TrimSpace(article.TextContent) != "" {
+			textContent = article.TextContent
+		}
+	}
+
+	if textContent == "" {
+		// Fallback: extract just the <body> tag using Go's HTML parser
+		s.logDebug("readability extraction failed, falling back to body tag extraction", "url", pageURL)
+		doc, htmlErr := html.Parse(bytes.NewReader(body))
+		if htmlErr != nil {
+			s.logError("failed to parse HTML", "error", htmlErr, "url", pageURL)
+			return "unable to parse the response", htmlErr
+		}
+		textContent = extractBodyContent(doc)
+	}
+
+	if strings.TrimSpace(textContent) == "" {
+		s.logWarn("extracted body content is empty", "url", pageURL)
+		return "fetched page contained no readable content", nil
+	}
+
+	// Perform recursive summarization
+	summary, err := s.summarizeContent(bot, textContent)
+	if err != nil {
+		s.logWarn("recursive summarization failed, falling back to raw content with warnings", "error", err)
+		return s.wrapSourceContentWithContext(textContent, matchedResult, llmContext), nil
+	}
+
```

Comment:
If we move forward with the `isSummaryUnsafe`, we can add it here too:

```suggestion
	if unsafe, reason := isSummaryUnsafe(summary); unsafe {
		s.logWarn("summarized content rejected", "reason", reason)
		return s.wrapSourceContentWithContext(summary, matchedResult, llmContext), nil
	}
	
```